src/lib.rs

```diff
@@ -105,23 +99,64 @@ impl IO<()> {
 }
 
 /// Various combinators on the IO monad.
-impl<O> IO<O> {
+impl<'a, O> IO<'a, O>
+where
+    O: 'a
+{
+
+    pub (crate) fn new (
+        action: impl Action<'a, Output = O> + 'a
+    ) -> IO<'a, O> {
+        IO (Eval (Box::new(action)))
+    }
+
+    /// Make it easier to define new relatively simple actions
+    /// without all the boilerplate.
+    pub (crate) fn from_fn (
+        f: impl FnOnce (Context<'_>) -> Result<O> + 'a
+    ) -> IO<'a, O> {
+        struct FromFn <'x, F, O> (Option<F>, PhantomData<&'x O>);
+
+        impl<'x,F, O> Action<'x> for FromFn<'x, F, O>
+        where
+            F: FnOnce (Context<'_>) -> Result<O> + 'x,
+            O: 'x,
+        {
+            type Output = O;
+
+            fn perform (&mut self, cx: Context<'_>) -> Result<Self::Output> {
+                let FromFn (f, _) = self;
+
+                // Ensure this is the first time the action is called.
+                // This is required because `F` is `FnOnce` and not `FnMut`
+                // and the borrow checker will complain about moving it if
+                // we don't use `take` or something similar.
+                let f = f.take().expect("Re-evaluation of IO action!");
+
+                f(cx)
+            }
+        }
+
+        IO::new(FromFn (Some (f), PhantomData))
+    }
 
     /// Perform a monadic binding operation. For the IO monad, this
     /// means simply sequencing `self` to happen before calling `f`
     /// on the result and then performing the value returned from
     /// `f`.
-    pub fn bind <R> (self, f: impl FnOnce (O) -> IO<R> + 'static) -> IO<R>
+    pub fn bind <R> (self, f: impl FnOnce (O) -> IO<'a, R> + 'a) -> IO<'a, R>
```

Comment:
I've added some additional explanation, let me know if it helps :)

Reply:
This might be out of scope for the docs here, but I think maybe showing the function signature for map may be useful as a reminded/explanation of what the difference between map and bind/and_then is? The docs for this function are already aimed at a "FP baby" like me, since anyone familiar with FP will already know what bind is.

Other than that the and_then comparison is really good and completely clears up what this does completely.